guides/micronaut-metrics/micronaut-metrics.adoc

```diff
@@ -0,0 +1,309 @@
+callout:mapped-entity-id[2]
+callout:generated-value[3]
+
+=== BookRepository
+
+Next, create the `BookRepository` interface to define database operations. Micronaut Data will implement the interface at compilation time:
+
+source:BookRepository[]
+callout:jdbcrepository[1]
+<2> Specifies that `Book` is the root entity and the primary key type is `Long`.
+
+=== Data populator class
+
+Create a `DataPopulator` class to create some example database entries when the application starts:
+
+source:DataPopulator[]
+
+callout:singleton[1]
+callout:requires-not-env-test[2]
+callout:constructor-di[number=3,arg0=BookRepository]
+callout:event-listener[4]
+
+=== BookController
+
+Create `BookController` to access `Book` instances (and trigger JDBC metric data):
+
+source:BookController[]
+callout:controller[number=1,arg0=/books]
+callout:executes-on[2]
+callout:constructor-di[number=3,arg0=BookRepository]
+<4> Maps a `GET` request to `/books`, which returns a list of `Book`
+<5> Maps a `GET` request to `/books/{isbn}`, which attempts to show a `Book`. This illustrates the use of a URL path variable.
+
+=== Custom Metric
+To see custom metrics in action, create a service that periodically retrieves the current Bitcoin price in USD using REST.
+
+==== Model
+
+We'll need a data class to represent the REST response. Create the `BitcoinPrice` class:
+
+source:crypto/BitcoinPrice[]
+
+callout:introspected[1]
+
+==== Kucoin Declarative HTTP Client
+
+Create declarative Micronaut HTTP Client interface that will be implemented at compile time:
+
+source:crypto/PriceClient[]
+
+callout:client[1]
+
+==== HTTP Service Configuration
+
+Configure the URLs for the Service ID `kucoin`.
+Modify `application.yml`
+
+resource:application.yml[tag=http-services]
+
+<1> Same ID we used with `@Client`
+<2> URL where the `books` service will reside.
+
+==== Service
+
+Create the `CryptoService` class which uses `PriceClient` and updates three custom meters:
+
+source:crypto/CryptoService[]
+callout:singleton[1]
+<2> Use constructor injection to inject `PriceClient`, `MeterRegistry`, and `ObjectMapper` beans
+<3> Create a `Counter` to track the total number of updates
+<4> Create a `Timer` to track the total time performing updates
+<5> Create a `Gauge` to track the most recent update
+<6> Use the `Scheduled` annotation to configure regular updates
+<7> Update the `Timer`
+<8> Increment the `Counter`
+<9> Update the `Gauge` with the latest price
+
+== Testing the Application
+
+Create a configuration file for tests `src/test/resources/application-test.yml`.
+
+testResource:application-test.yml[]
+
+<1> Configure the default datasource to use the MySQL database provided by TestContainers for the test environment. By using a specially modified JDBC URL, Testcontainers provides a disposable stand-in database that can be used without requiring modification to your application code. See https://www.testcontainers.org/modules/databases/jdbc/[Test Containers JDBC URL].
```

Comment:
Should Testcontainers have the "C" capitalized? It looks like only the "T" should be capitalized. 

https://github.com/testcontainers/testcontainers-java 

Whichever one you pick, I'd do a quick search and make sure you're consistent throughout.